### ISSUE TITLE
Remove Customer ID from subscription model

### ISSUE THREAD

**Issue by srichmond** [workfloworchestrator/orchestrator-core#238]

**Comment by Sparrow1029** [comment]
Looking for clarification on this story regarding AC #2: "A consumer of orch-core could have a new table with an optional schema migration that is not part of the core"

Is this to say that once the `consumer_id` column is removed from the schema, that any current consumer of orch-core that uses that field will have to create a migration to ensure their code does not break?

**Comment by pboers1988** [comment]
Correct. This is a breaking change and would require users to carefully migratie existing information to a new database table.

**Comment by Sparrow1029** [comment]
Circling back to this after a busy sprint...

I wonder about the decision to eventually remove the `customer_id` column entirely from the `Subscription` model.

There is almost always an entity (organization, customer, "owner") associated with a Subscription in the Service/Product/Resource BSS model. I believe that removing this column entirely might pointlessly make the orchestrator-core library _much_ harder to extend/use for downstream consumers.

If we remove the `customer_id`, a user of the core library will probably have the use case to associate a customer to a Subscription—whether it be one-many, many-many, many-one—and will have to extend/override the included `SubscriptionTable` ORM model in `orchestrator/db/models.py` to build the appropriate queries joined to their own orchestrator-db `Customer` (or just use the column value to query an external resource).

Instead of removing this column, maybe it can be changed to a `VARCHAR` type to allow flexibility for the end user without removing a core functionality of the orchestrator. By converting it to a string type, a user could associate whatever foreign key best suits their use-case, be it an LDAP username, Integer key, Organization name, or UUID. Then, if a user decides to implement a `SubscriptionCustomer` table, they can make the necessary associations however they choose, without having to sub-class or create a new `SubscriptionTable` ORM model.

I could also be completely missing the intent of this issue, or have a misunderstanding of how the new models would be implemented... welcome feedback on these thoughts!

**Comment by Mark90** [comment]
I'm also in favor of removing the column from the core Subscription model.

Another option instead of moving customer_id to a new (implementation specific) SQL table, is to make it part of the generic subscription metadata being implemented in #266.

**Comment by Sparrow1029** [comment]
> Another option instead of moving customer_id to a new (implementation specific) SQL table, is to make it part of the generic subscription metadata being implemented in #266. 

This is an acceptable compromise -- still a breaking change that can require some significant refactoring for downstream users, but if the column has to go...

**Comment by hanstrompert** [comment]
> > Another option instead of moving customer_id to a new (implementation specific) SQL table, is to make it part of the generic subscription metadata being implemented in #266.
> 
> This is an acceptable compromise -- still a breaking change that can require some significant refactoring for downstream users, but if the column has to go...

Subscription ownership is a key feature in the orchestrator, and is not something optional. Searching or filtering on the customer name (and for SURF also the customer abbreviation) is an important feature in the orchestrator client. This makes integrating a CRM that contains this information highly recommended. For me, meta data is optional, customer description (and probably abbreviation) is not. This does not mean that the orchestrator client does not work when you do not integrate a CRM but then you only see and can search or filter on the internal customer ID as stored in the orchestrator. The only thing we wanted to do here is to lift the restriction for a customer ID to be a UUID by introducing a linking table.

**Comment by pboers1988** [comment]
Round up of the discusion:

- [ ] We need to loosen the restriction on the customer_id column to be a varchar. 
- [ ] Furthermore we need to make an example implementation where the following can be set as default:
  - [ ] Fullname
  - [ ] Shortcode
  - [ ] Identifier

**Comment by Mark90** [comment]
> [ ] Add configuration option DEFAULT_CUSTOMER_ONLY that will enable GraphQL endpoint /customers that serves the one and only default customer as specified above

As discussed in the meeting today, it seems to make more sense to use the existing `/api/graphql` endpoint and add a `customers` query to the strawberry `Query` which (conditionally) returns the default customer
